### PR TITLE
Improve compatibility with plugin Delivery & Pickup Date Time for WooCommerce

### PR DESCRIPTION
**Pull request by vlad-dedikov**

### CODE REVIEW

File: inc/compat/plugins/compat-plugin-woo-delivery.php [hunk 1]
```diff
@@ -165,15 +239,15 @@ public function maybe_set_step_incomplete( $is_step_complete ) {
...
-		$order_type = FluidCheckout_Steps::instance()->get_checkout_field_value_from_session( 'coderockz_woo_delivery_delivery_selection_box' );
-		$delivery_date = FluidCheckout_Steps::instance()->get_checkout_field_value_from_session( 'coderockz_woo_delivery_date_field' );
-		$delivery_time = FluidCheckout_Steps::instance()->get_checkout_field_value_from_session( 'coderockz_woo_delivery_time_field' );
-		$pickup_date = FluidCheckout_Steps::instance()->get_checkout_field_value_from_session( 'coderockz_woo_delivery_pickup_date_field' );
-		$pickup_time = FluidCheckout_Steps::instance()->get_checkout_field_value_from_session( 'coderockz_woo_delivery_pickup_time_field' );
+		$order_type = WC()->session->get( 'coderockz_woo_delivery_delivery_selection_box' );
+		$delivery_date = WC()->session->get( 'coderockz_woo_delivery_date_field' );
+		$delivery_time = WC()->session->get( 'coderockz_woo_delivery_time_field' );
+		$pickup_date = WC()->session->get( 'coderockz_woo_delivery_pickup_date_field' );
+		$pickup_time = WC()->session->get( 'coderockz_woo_delivery_pickup_time_field' );
```
[the remark this answers is not in the document]
Comment (diegoversiani) [reply inc/compat/plugins/compat-plugin-woo-delivery.php:296]: @vlad-dedikov I read your comment on the Clickup task, but it is still not clear why it did not work with the functions from Fluid Checkout.

1. Was it not working because the Coderockz plugin uses the session keys without the prefix `fc_`?
2. Could you describe the issues you had while trying to use the functions from Fluid Checkout?
3. If using the WooCommerce session directly is strictly needed, we should add a comment on the code for why that is the case so we do not refactor that in the future when we forget about these reasons.

**Review by diegoversiani — DISMISSED** [review]

File: inc/compat/plugins/compat-plugin-woo-delivery.php [hunk 2]
```diff
@@ -225,11 +299,11 @@ public function add_substep_text_lines_delivery_date( $review_text_lines = array
...
```
[the remark this answers is not in the document]
Comment (diegoversiani) [reply inc/compat/plugins/compat-plugin-woo-delivery.php:365]: See comment https://github.com/fluid-checkout/fluid-checkout/pull/451#discussion_r1869375597

File: inc/compat/plugins/compat-plugin-woo-delivery.php [hunk 3]
```diff
@@ -148,6 +167,61 @@ public function output_substep_delivery_options( $step_id ) {
...
+		$fields = array(
+			'coderockz_woo_delivery_date_field',
+			'coderockz_woo_delivery_time_field',
+			'coderockz_woo_delivery_pickup_date_field',
+			'coderockz_woo_delivery_pickup_time_field'
+		);
+
+		foreach ( $fields as $field ) {
+			if ( array_key_exists( $field, $posted_data ) && ! empty( $posted_data[ $field ] ) ) {
+				WC()->session->set( $field, $posted_data[ $field ] );
```
[the remark this answers is not in the document]
Comment (diegoversiani) [reply inc/compat/plugins/compat-plugin-woo-delivery.php:234]: See comment https://github.com/fluid-checkout/fluid-checkout/pull/451#discussion_r1869375597

File: inc/compat/plugins/compat-plugin-woo-delivery.php [hunk 4]
```diff
@@ -43,8 +63,12 @@ public function late_hooks() {
 		$this->remove_action_for_class( 'woocommerce_review_order_before_payment', array( 'Coderockz_Woo_Delivery_Public', 'coderockz_woo_delivery_add_custom_field' ), 10 );
 		$this->remove_action_for_class( 'woocommerce_checkout_before_order_review_heading', array( 'Coderockz_Woo_Delivery_Public', 'coderockz_woo_delivery_add_custom_field' ), 10 );
 
-		// Get delivery date value from session
-		add_filter( 'woocommerce_checkout_get_value', array( $this, 'change_default_field_values_from_session' ), 200, 2 );
+		// TODO: CHECK WHY IT WAS REMOVED
+		// // Get delivery date value from session
+		// add_filter( 'woocommerce_checkout_get_value', array( $this, 'change_default_field_values_from_session' ), 200, 2 );
```
Comment (diegoversiani) [inline inc/compat/plugins/compat-plugin-woo-delivery.php:68]: @vlad-dedikov, why was this removed?

**Comment by diegoversiani** [comment]
@vlad-dedikov, I managed to merge the changes from Lite 4.0, but I'm not sure if your fixes are still working. The selected value is not being preserved when I refresh the checkout page.

File: inc/compat/plugins/compat-plugin-woo-delivery.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (vlad-dedikov) [reply inc/compat/plugins/compat-plugin-woo-delivery.php]: @diegoversiani I've search-replaced all of those and it appears to be working... I guess I already removed something that was clearing/resetting session values or it had something to do with JS.

File: inc/compat/plugins/compat-plugin-woo-delivery.php [hunk 5]
```diff
@@ -29,54 +38,63 @@ public function hooks() {
...
-			// Maybe set step as incomplete
-			add_filter( 'fc_is_step_complete_shipping', array( $this, 'maybe_set_step_incomplete' ), 10 );
+
+	/**
+	 * Add or remove hooks for the plugin elements.
+	 */
+	public function plugin_elements_hooks() {
+		// Bail if plugin classes and methods are not available
+		if ( ! class_exists( 'Coderockz_Woo_Delivery' ) || ! class_exists( 'Coderockz_Woo_Delivery_Public' ) ) { return; }
+		if ( ! method_exists( 'Coderockz_Woo_Delivery', 'get_plugin_name' ) || ! method_exists( 'Coderockz_Woo_Delivery', 'get_version' ) ) { return; }
```
[the remark this answers is not in the document]
Comment (vlad-dedikov) [reply inc/compat/plugins/compat-plugin-woo-delivery.php]: Those are removed.

File: inc/compat/plugins/compat-plugin-woo-delivery.php (hunk 4 above)
Comment (diegoversiani) [reply inc/compat/plugins/compat-plugin-woo-delivery.php:65]: @vlad-dedikov, maybe the function `maybe_set_datepicker_fields_session_values` is replacing what this code should be doing.

Check the behavior of the plugin and if it is working well we can remove this code entirely.

Comment (diegoversiani) [reply inc/compat/plugins/compat-plugin-woo-delivery.php]: @vlad-dedikov, I restored this part of the code since it does not seem to have any negative effects.
